### ISSUE TITLE
potential data races

### ISSUE THREAD

**Issue by 191830230** [rclone/rclone#6962]
Hello developer, I used a static data race detector to check this project and found suspicious data race defects. I have done preliminary filtering and categorized the defects, choosing only one from each category. However, due to my unfamiliarity with the overall architecture of the project, there may still be false positives after filtering. Below are the potential data race issues I found(not all stack traces are listed)
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Potential race condition:
 Access1:
        abort = true
        ^
rclone/vfs/read.go:240:3 ->
rclone/vfs/read.go:234:4
 
 Access2:
        for *poff != off && !abort {
                             ^
rclone/vfs/read.go:241:23 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Potential race condition:
 Access1:
        err = s.serveHTTP()
        ^
rclone/cmd/serve/dlna/dlna.go:310:3 ->
rclone/cmd/serve/dlna/dlna.go:306:3
 
 Access2:
        return nil
        ^
rclone/cmd/serve/dlna/dlna.go:312:2 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Potential race condition:
 Access1:
        if err != nil {
           ^
rclone/cmd/serve/dlna/dlna.go:310:3 ->
rclone/cmd/serve/dlna/dlna.go:307:6
 
 Access2:
        return nil
        ^
rclone/cmd/serve/dlna/dlna.go:312:2 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Potential race condition:
 Access1:
        fs.Logf(s.f, "Error on serving HTTP server: %v", err)
                                                         ^
rclone/cmd/serve/dlna/dlna.go:310:3 ->
rclone/cmd/serve/dlna/dlna.go:308:53
 
 Access2:
        return nil
        ^
rclone/cmd/serve/dlna/dlna.go:312:2

**Comment by ncw** [comment]
> Hello developer, I used a static data race detector to check this project and found suspicious data race defects.

What tool were you using?

> I have done preliminary filtering and categorized the defects, choosing only one from each category. However, due to my unfamiliarity with the overall architecture of the project, there may still be false positives after filtering. Below are the potential data race issues I found(not all stack traces are listed)

> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> Potential race condition:
>  Access1:
>         abort = true
>         ^
> rclone/vfs/read.go:240:3 ->
> rclone/vfs/read.go:234:4
>  
>  Access2:
>         for *poff != off && !abort {
>                              ^
> rclone/vfs/read.go:241:23 

This is a potential race. However to trigger it the timeout would have to expire before the `cond.Wait()` is run which is very unlikely.

https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/vfs/read.go#L225-L243

Not sure the best way to fix that. Maybe turn `abort` into an `atomic` reference? What do you think.

> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> Potential race condition:
>  Access1:
>         err = s.serveHTTP()
>         ^
> rclone/cmd/serve/dlna/dlna.go:310:3 ->
> rclone/cmd/serve/dlna/dlna.go:306:3
>  
>  Access2:
>         return nil
>         ^
> rclone/cmd/serve/dlna/dlna.go:312:2 

Yes I think this is a race condition. Subtle though as it depends on the named return variable `err`.

https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/cmd/serve/dlna/dlna.go#L284

https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/cmd/serve/dlna/dlna.go#L306-L312

This should fix it - would you agree?

```diff
--- a/cmd/serve/dlna/dlna.go
+++ b/cmd/serve/dlna/dlna.go
@@ -303,7 +303,7 @@ func (s *server) Serve() (err error) {
 	go func() {
 		fs.Logf(s.f, "Serving HTTP on %s", s.HTTPConn.Addr().String())
 
-		err = s.serveHTTP()
+		err := s.serveHTTP()
 		if err != nil {
 			fs.Logf(s.f, "Error on serving HTTP server: %v", err)
 		}
```

> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> Potential race condition:
>  Access1:
>         if err != nil {
>            ^
> rclone/cmd/serve/dlna/dlna.go:310:3 ->
> rclone/cmd/serve/dlna/dlna.go:307:6
>  
>  Access2:
>         return nil
>         ^
> rclone/cmd/serve/dlna/dlna.go:312:2 

I think this is the same problem as the one above with the same fix.

> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> Potential race condition:
>  Access1:
>         fs.Logf(s.f, "Error on serving HTTP server: %v", err)
>                                                          ^
> rclone/cmd/serve/dlna/dlna.go:310:3 ->
> rclone/cmd/serve/dlna/dlna.go:308:53
>  
>  Access2:
>         return nil
>         ^
> rclone/cmd/serve/dlna/dlna.go:312:2 

I think this is the same problem as the one above with the same fix also.

**Comment by 191830230** [comment]
&nbsp;
&nbsp;
------------------&nbsp;Original&nbsp;------------------
From: ***@***.***&gt;;
Date: &nbsp;Thu, Apr 20, 2023 04:54 PM
To: ***@***.***&gt;; 
Cc: ***@***.***&gt;; ***@***.***&gt;; 
Subject: &nbsp;Re: [rclone/rclone] potential data races (Issue #6962)

&nbsp;



  
Hello developer, I used a static data race detector to check this project and found suspicious data race defects.
  
What tool were you using?

This tool was designed by myself, I will make it public later.

  
I have done preliminary filtering and categorized the defects, choosing only one from each category. However, due to my unfamiliarity with the overall architecture of the project, there may still be false positives after filtering. Below are the potential data race issues I found(not all stack traces are listed)
   
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Potential race condition:
 Access1:
 abort = true
 ^
 rclone/vfs/read.go:240:3 -&gt;
 rclone/vfs/read.go:234:4
 
Access2:
 for *poff != off &amp;&amp; !abort {
 ^
 rclone/vfs/read.go:241:23
  
This is a potential race. However to trigger it the timeout would have to expire before the cond.Wait() is run which is very unlikely.
 
https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/vfs/read.go#L225-L243
 
Not sure the best way to fix that. Maybe turn abort into an atomic reference? What do you think.

I think atomic reference is a good fix.

  
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Potential race condition:
 Access1:
 err = s.serveHTTP()
 ^
 rclone/cmd/serve/dlna/dlna.go:310:3 -&gt;
 rclone/cmd/serve/dlna/dlna.go:306:3
 
Access2:
 return nil
 ^
 rclone/cmd/serve/dlna/dlna.go:312:2
  
Yes I think this is a race condition. Subtle though as it depends on the named return variable err.
 
https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/cmd/serve/dlna/dlna.go#L284
 
https://github.com/rclone/rclone/blob/589b7b4873f8ba1d77f982ce14d4ac9a8adb18f6/cmd/serve/dlna/dlna.go#L306-L312
 
This should fix it - would you agree?

Yes, I agree.

 --- a/cmd/serve/dlna/dlna.go +++ b/cmd/serve/dlna/dlna.go @@ -303,7 +303,7 @@ func (s *server) Serve() (err error) {  	go func() {  		fs.Logf(s.f, "Serving HTTP on %s", s.HTTPConn.Addr().String())   -		err = s.serveHTTP() +		err := s.serveHTTP()  		if err != nil {  			fs.Logf(s.f, "Error on serving HTTP server: %v", err)  		}
  
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Potential race condition:
 Access1:
 if err != nil {
 ^
 rclone/cmd/serve/dlna/dlna.go:310:3 -&gt;
 rclone/cmd/serve/dlna/dlna.go:307:6
 
Access2:
 return nil
 ^
 rclone/cmd/serve/dlna/dlna.go:312:2
  
I think this is the same problem as the one above with the same fix.
  
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Potential race condition:
 Access1:
 fs.Logf(s.f, "Error on serving HTTP server: %v", err)
 ^
 rclone/cmd/serve/dlna/dlna.go:310:3 -&gt;
 rclone/cmd/serve/dlna/dlna.go:308:53
 
Access2:
 return nil
 ^
 rclone/cmd/serve/dlna/dlna.go:312:2
  
I think this is the same problem as the one above with the same fix also.

**Comment by ncw** [comment]
I've merged the fixes for this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.63
